mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala

```diff
@@ -130,9 +131,54 @@ class LinearRegression(override val uid: String)
   def setWeightCol(value: String): this.type = set(weightCol, value)
   setDefault(weightCol -> "")
 
+  /**
+   * Set the solver algorithm used for optimization.
+   * In case of linear regression, this can be "l-bfgs", "normal" and "auto".
+   * The default value is "auto" which means that the solver algorithm is
+   * selected automatically.
+   * @group setParam
+   */
+  def setSolver(value: String): this.type = set(solver, value)
+  setDefault(solver -> "auto")
+
   override protected def train(dataset: DataFrame): LinearRegressionModel = {
+    // Extract the number of features before deciding optimization solver.
+    val numFeatures = dataset.select(col($(featuresCol))).limit(1).map {
+      case Row(features: Vector) =>
+        features.size
+    }.toArray()(0)
     // Extract columns from data.  If dataset is persisted, do not persist instances.
```

Comment:
Remove `// Extract columns from data.  If dataset is persisted, do not persist instances.`